### ISSUE TITLE
Support for rectangular matrices in the matrix library

### ISSUE THREAD

**Issue by oskgo** [EasyCrypt/easycrypt#95]
The current matrix library only supports square matrices. I want to extend this to arbitrary shapes.

First, to be clear, I'm not asking anyone else to do this. I just want to make sure that what I'm doing is useful.

I have a few ideas for the design and was wondering which one (if any) aligns with what EasyCrypt wants.

- A subtype of lists of lists where all the inner lists have the same length. This avoids the need to carry around extra information for the shape of the matrix, and allows defining vectors without subtyping.
- A subtype of functions annotated with the shape of the matrix `(int -> int -> t) * int * int`. This is closer to the original implementation.

If there is work being done already and you could use some help I would be happy to join.

**Comment by strub** [comment]
Hi,

This is clearly something we are interested in and that we planed to do at some point, so welcome aboard!

To answer your first question, I would go with the `(int -> int -> t) * int * int` solution as this will allow you to reuse a large part of the current library and is, IMO, far simpler to deal with. Eventually, square matrices will be a refinement of matrices.

There is nobody working on that currently. However, we planed to use such a library to test @AntoineSere's work on subtypes & type-classes. I think that it is important that your work integrates well with Antoine's work so that the final merge goes smoothly. Maybe can we have a discussion at some point?
